app/account/page.jsx

```diff
@@ -46,29 +48,68 @@ function a11yProps(index) {
 
 export default function BasicTabs() {
   const [currentTab, setCurrentTab] = React.useState(0);
+  const [purchases, setPurchases] = React.useState([]);
   const { user, userData, setUserData } = useUserContext();
   const router = useRouter();
+  let groupedByDate;
+  const days = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
+  const months = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];
 
-  const fetchUserPath = '../api/userData';
-  const action = 'getuser';
-  const collection = 'userdata';
-  const document = user?.uid.toString() || null;
-  const requestPath = `${fetchUserPath}?action=${action}&collection=${collection}&document=${document}`
-  const { data, error, isLoading } = useSWR(requestPath, customGetter);
   useEffect(() => {
-
     if (user === null) {
-      router.push('/signin')
-    }
+      router.push('/signin');
+    };
+  });
+
+  const fetchUserPath = '../api/userData';
+  const fetchUserAction = 'getuser';
+  const fetchUserCollection = 'userdata';
+  const fetchUserDocument = user?.uid.toString() || null;
+  const fetchUserRequestPath = `${fetchUserPath}?action=${fetchUserAction}&collection=${fetchUserCollection}&document=${fetchUserDocument}`;
+
+  const purchasesPath = '../api/userData';
+  const purchasesAction = 'getpurchases';
+  const purchasesCollection = 'purchases';
+  const purchasesDocumentKey = user?.uid.toString() || null;
+  const purchasesRequestPath = `${purchasesPath}?action=${purchasesAction}&collection=${purchasesCollection}&document=${purchasesDocumentKey}`;
+
+  const accountFetchKeys = {
+    fetchUserRequestPath,
+    purchasesRequestPath,
+  };
+
+  const getAccountData = async (accountFetchKeys) => {
+    const { fetchUserRequestPath, purchasesRequestPath } = accountFetchKeys;
+    const userAccountData = await customGetter(fetchUserRequestPath);
+    const pastPurchases = await customGetter(purchasesRequestPath);
+    return { userAccountData, pastPurchases };
+  };
+
+  const { data, error, isLoading } = useSWR(accountFetchKeys, getAccountData);
+
+  useEffect(() => {
     if (!isLoading) {
-      setUserData(data)
-    }
-  })
+      const { userAccountData, pastPurchases } = data;
+      setUserData(userAccountData);
+      setPurchases(pastPurchases);
+    };
+  }, [data, isLoading, setUserData, setPurchases, router, user]);
 
   const handleChange = (event, newValue) => {
     setCurrentTab(newValue);
   };
 
+  if (purchases.length) {
+    groupedByDate = purchases.reduce((result, item) => {
```

Comment:
so this groupedByDate var is being created but not fully initialized earlier in the component. but here you set it's value. curious, should this be handled in a state setter/getter? Or since it's only calculated one time, it's done this way?

Reply:
thats an interesting question. As its value is only set once and cannot be changed without leaving the page (you would have to go make a purchase and then go back to see changes), I think that setting it as a regular JS variable is ok. However I also feel like the way that it is initialized as 'undefined' looks kind of awkward. I will probably initialize it as an empty object right before the value is set